### PR TITLE
Context propagation refactorings

### PR DESCRIPTION
**Pull request by JonasKunz**
## What does this PR do?

Move the following responsibilites from `Span` to `ElasticContext`:
 * Creation of child spans
 * Context propagation

In addition to make plugins simpler I decided to ensure that `ElasticContext` is never null by returning an empty context instead. This avoids a lot of null-checks in the plugins.

Part of #3004 . This PR also changes that context propagation always happens, even if there is no active span.
E.g. if someone just updated the baggage, we want to still create baggage headers, even if no trace-context will be propagated.

This PR is a refactoring and should NOT change any actual functionality of the agent.

## Checklist

- [x] This is something else
  - [ ] ~I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/main/CHANGELOG.asciidoc)~

### CODE REVIEW

File: apm-agent-core/src/main/java/co/elastic/apm/agent/impl/ActiveStack.java [hunk 1]
```diff
@@ -80,7 +80,7 @@ public ElasticContext<?> currentContext() {
         if (current instanceof ElasticContextWrapper) {
             return ((ElasticContextWrapper<?>) current).getWrappedContext();
         }
-        return current;
+        return current != null ? current : EmptyElasticContext.INSTANCE;
```
Comment (JonasKunz) [inline apm-agent-core/src/main/java/co/elastic/apm/agent/impl/ActiveStack.java:83]: This change has been to simplify plugins: Instead of returning null when there is no context, we return an empty context. This avoids a lot of null-checks in the plugins.
If one really needs to know wheter a context is active or not, it is possible to do so by calling `ElasticContext.isEmpty()`. This is for example useful to optimize intra-process context propagation: No need to propagate empty contexts here.

File: apm-agent-tracer/src/main/java/co/elastic/apm/agent/tracer/ElasticContext.java [hunk 2]
```diff
@@ -18,26 +18,113 @@
...
 
-public interface ElasticContext<T extends ElasticContext<T>> {
+import co.elastic.apm.agent.tracer.dispatch.BinaryHeaderSetter;
+import co.elastic.apm.agent.tracer.dispatch.TextHeaderGetter;
+import co.elastic.apm.agent.tracer.dispatch.TextHeaderSetter;
+import co.elastic.apm.agent.tracer.reference.ReferenceCounted;
+
+import javax.annotation.Nullable;
+
+public interface ElasticContext<T extends ElasticContext<T>> extends ReferenceCounted, Activateable<T> {
```
Comment (JonasKunz) [inline apm-agent-tracer/src/main/java/co/elastic/apm/agent/tracer/ElasticContext.java:28]: `ElasticContext` is now reference counted. The reason is that context can reference spans and spans are reference counted.
E.g. if you were to propagate a context to another thread, you need to make sure to do an `incrementReferenceCount` on the context, just like we did for `Spans` until now.

File: apm-agent-core/src/main/java/co/elastic/apm/agent/impl/ActiveStack.java [hunk 3]
```diff
@@ -122,22 +122,15 @@ boolean deactivate(ElasticContext<?> context, List<ActivationListener> activatio
         ElasticContext<?> activeContext = currentContext();
         activeContextStack.remove();
 
-        AbstractSpan<?> span = context.getSpan();
-
-        if (activeContext != context && activeContext instanceof ElasticContextWrapper) {
```
Comment (JonasKunz) [inline apm-agent-core/src/main/java/co/elastic/apm/agent/impl/ActiveStack.java:127]: This unwrapping was unnecessary: `currentContext` has already been doing the unwrapping.

File: apm-agent-plugins/apm-apache-httpclient/apm-apache-httpclient3-plugin/src/main/java/co/elastic/apm/agent/httpclient/v3/HttpClient3Instrumentation.java [hunk 4]
```diff
@@ -82,54 +80,46 @@ public static class HttpClient3Advice {
...
-            try {
-                httpClientURI = httpMethod.getURI();
-                host = hostConfiguration.getHost();
-                port = hostConfiguration.getPort();
-                protocol = hostConfiguration.getProtocol().getScheme();
-                uri = httpClientURI.toString();
-            } catch (Exception e) {
+            Span<?> span = null;
+            final ElasticContext<?> parent = TracerAwareInstrumentation.tracer.currentContext();
+            if (parent.getSpan() != null) {
```
Comment (SylvainJuge) [inline apm-agent-plugins/apm-apache-httpclient/apm-apache-httpclient3-plugin/src/main/java/co/elastic/apm/agent/httpclient/v3/HttpClient3Instrumentation.java:85]: [nitpicking] keeping the shortcut return when there is no span avoids one level of if nesting.

File: apm-agent-plugins/apm-apache-httpclient/apm-apache-httpclient4-plugin/src/main/java/co/elastic/apm/agent/httpclient/v4/ApacheHttpAsyncClientInstrumentation.java [hunk 5]
```diff
@@ -89,42 +89,37 @@ public static class ApacheHttpAsyncClientAdvice {
...
+            Span<?> span = enter != null ? (Span<?>) enter[2] : null;
             if (span != null) {
                 // Deactivate in this thread. Span will be ended and reported by the listener
                 span.deactivate();
-
-                if (!((Boolean) enter[2])) {
-                    // Listener is not wrapped- we need to end the span so to avoid leak and report error if occurred during method invocation
-                    span.captureException(t);
-                    span.end();
-                }
```
Comment (SylvainJuge) [inline apm-agent-plugins/apm-apache-httpclient/apm-apache-httpclient4-plugin/src/main/java/co/elastic/apm/agent/httpclient/v4/ApacheHttpAsyncClientInstrumentation.java:127]: instead of removing this we should probably test for `t != null` if an exception was thrown in the instrumented method.

File: apm-agent-plugins/apm-apache-httpclient/apm-apache-httpclient3-plugin/src/main/java/co/elastic/apm/agent/httpclient/v3/HttpClient3Instrumentation.java (hunk 4 above)
Comment (JonasKunz) [reply apm-agent-plugins/apm-apache-httpclient/apm-apache-httpclient3-plugin/src/main/java/co/elastic/apm/agent/httpclient/v3/HttpClient3Instrumentation.java:85]: Sorry, should have mentioned it in the PR description (updated it now):
This PR also changes that context propagation always happens, even if there is no active span.
E.g. if someone just updated the baggage, we want to still create baggage headers, even if no trace-context will be propagated. Baggage is conceptually independent from tracing.

For this reason I had to remove the early exits in order to make sure that the context propagation part is always reached.

File: apm-agent-tracer/src/main/java/co/elastic/apm/agent/tracer/ElasticContext.java [hunk 6]
```diff
@@ -18,26 +18,113 @@
...
+
+    /**
+     * @return the span/transaction that is associated to this context, {@literal null} if there is none
+     */
+    @Nullable
+    AbstractSpan<?> getSpan();
```
[the remark this answers is not in the document]
Comment (JonasKunz) [reply apm-agent-tracer/src/main/java/co/elastic/apm/agent/tracer/ElasticContext.java:34]: I would find `ElasticContext.getActive()` also confusing: Does it return the active span, baggage, or both?
What about `getAbstractSpan`? We already have the `AbstractSpan` class in the API indicating that it can be either a transaction or a span?

File: apm-agent-plugins/apm-apache-httpclient/apm-apache-httpclient4-plugin/src/main/java/co/elastic/apm/agent/httpclient/v4/LegacyApacheHttpClientInstrumentation.java [hunk 7]
```diff
@@ -81,40 +79,32 @@ public static class LegacyApacheHttpClientAdvice {
...
-                uri = uriRequest.getURI();
-            } else {
-                RequestLine requestLine = request.getRequestLine();
-                method = requestLine.getMethod();
-                try {
-                    uri = new URI(requestLine.getUri());
-                } catch (URISyntaxException ignore) {
+            final ElasticContext<?> parent = tracer.currentContext();
+            Span<?> span = null;
+            if (parent.getSpan() != null) {
```
[the remark this answers is not in the document]
Comment (JonasKunz) [reply apm-agent-plugins/apm-apache-httpclient/apm-apache-httpclient4-plugin/src/main/java/co/elastic/apm/agent/httpclient/v4/LegacyApacheHttpClientInstrumentation.java:84]: The previous code also checked for the presence of an active `AbstractSpan`, just like the changed code?
The presence of an `AbstractSpan` implies the presence of a `Transaction`.
Normally this check would not even be necessary (due to non-null `ElasticContext`s), but it was kept here to potentially avoid the allocations below.

File: apm-agent-plugins/apm-apache-httpclient/apm-apache-httpclient4-plugin/src/main/java/co/elastic/apm/agent/httpclient/v4/ApacheHttpAsyncClientInstrumentation.java [hunk 8]
```diff
@@ -89,42 +89,37 @@ public static class ApacheHttpAsyncClientAdvice {
...
-
-                wrappedProducer = asyncHelper.wrapRequestProducer(requestProducer, span, null);
-                wrappedFutureCallback = asyncHelper.wrapFutureCallback(futureCallback, context, span);
-                responseFutureWrapped = true;
-            } else {
-                wrappedProducer = asyncHelper.wrapRequestProducer(requestProducer, null, parent);
+            Span<?> span = null;
+            ElasticContext<?> parent = tracer.currentContext();
+            if (parent.getSpan() != null) {
+                span = parent.createExitSpan();
```
[the remark this answers is not in the document]
Comment (JonasKunz) [reply apm-agent-plugins/apm-apache-httpclient/apm-apache-httpclient4-plugin/src/main/java/co/elastic/apm/agent/httpclient/v4/ApacheHttpAsyncClientInstrumentation.java:103]: This is an example where the check is not even necessary anymore due to contexts being non-null, I'll remove it.

File: apm-agent-plugins/apm-apache-httpclient/apm-apache-httpclient4-plugin/src/main/java/co/elastic/apm/agent/httpclient/v4/ApacheHttpClientInstrumentation.java [hunk 9]
```diff
@@ -54,25 +53,17 @@ public static class ApacheHttpClientAdvice {
         @Advice.OnMethodEnter(suppress = Throwable.class, inline = false)
         public static Object onBeforeExecute(@Advice.Argument(0) HttpRoute route,
                                              @Advice.Argument(1) HttpRequestWrapper request) {
-            AbstractSpan<?> parent = tracer.getActive();
-            if (parent == null) {
-                return null;
-            }
```
[the remark this answers is not in the document]
Comment (JonasKunz) [reply apm-agent-plugins/apm-apache-httpclient/apm-apache-httpclient4-plugin/src/main/java/co/elastic/apm/agent/httpclient/v4/ApacheHttpClientInstrumentation.java:60]: See https://github.com/elastic/apm-agent-java/pull/3206#discussion_r1242185270, the early-out had to be removed to enforce that context propagation always happens.

File: apm-agent-plugins/apm-apache-httpclient/apm-apache-httpclient4-plugin/src/main/java/co/elastic/apm/agent/httpclient/v4/ApacheHttpClientInstrumentation.java [hunk 10]
```diff
@@ -54,25 +53,17 @@ public static class ApacheHttpClientAdvice {
...
+            if (parent.getSpan() != null) {
+                span = HttpClientHelper.startHttpClientSpan(parent, request.getMethod(), request.getURI(), route.getTargetHost().getHostName());
 
-            if (!HeaderUtils.containsAny(tracer.getTraceHeaderNames(), request, RequestHeaderAccessor.INSTANCE)) {
                 if (span != null) {
-                    span.propagateTraceContext(request, RequestHeaderAccessor.INSTANCE);
-                } else if (!HeaderUtils.containsAny(tracer.getTraceHeaderNames(), request, RequestHeaderAccessor.INSTANCE)) {
-                    // re-adds the header on redirects
-                    parent.propagateTraceContext(request, RequestHeaderAccessor.INSTANCE);
+                    span.activate();
```
[the remark this answers is not in the document]
Comment (JonasKunz) [reply apm-agent-plugins/apm-apache-httpclient/apm-apache-httpclient4-plugin/src/main/java/co/elastic/apm/agent/httpclient/v4/ApacheHttpClientInstrumentation.java:62]: Sorry, I don't get what you mean here? Whenever the exit-span creation succeeds, it will also be activated?

File: apm-agent-plugins/apm-apache-httpclient/apm-apache-httpclient4-plugin/src/main/java/co/elastic/apm/agent/httpclient/v4/LegacyApacheHttpClientInstrumentation.java [hunk 11]
```diff
@@ -81,40 +79,32 @@ public static class LegacyApacheHttpClientAdvice {
...
                 if (span != null) {
-                    span.propagateTraceContext(request, RequestHeaderAccessor.INSTANCE);
-                } else if (!HeaderUtils.containsAny(tracer.getTraceHeaderNames(), request, RequestHeaderAccessor.INSTANCE)) {
-                    // re-adds the header on redirects
-                    parent.propagateTraceContext(request, RequestHeaderAccessor.INSTANCE);
+                    span.activate();
                 }
             }
 
+            tracer.currentContext().propagateContext(request, RequestHeaderAccessor.INSTANCE, RequestHeaderAccessor.INSTANCE);
```
[the remark this answers is not in the document]
Comment (JonasKunz) [reply apm-agent-plugins/apm-apache-httpclient/apm-apache-httpclient4-plugin/src/main/java/co/elastic/apm/agent/httpclient/v4/LegacyApacheHttpClientInstrumentation.java:107]: See https://github.com/elastic/apm-agent-java/pull/3206#discussion_r1242185270

File: apm-agent-plugins/apm-apache-httpclient/apm-apache-httpclient4-plugin/src/main/java/co/elastic/apm/agent/httpclient/v4/ApacheHttpAsyncClientInstrumentation.java (hunk 5 above)
Comment (SylvainJuge) [reply apm-agent-plugins/apm-apache-httpclient/apm-apache-httpclient4-plugin/src/main/java/co/elastic/apm/agent/httpclient/v4/ApacheHttpAsyncClientInstrumentation.java:127]: Good idea to add a test, checking this kind of corner cases with coverage is usually helpful !

File: apm-agent-tracer/src/main/java/co/elastic/apm/agent/tracer/ElasticContext.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (SylvainJuge) [reply apm-agent-tracer/src/main/java/co/elastic/apm/agent/tracer/ElasticContext.java:34]: Let's take `getAbstractSpan` if you think that's clearer, the return type also helps to understand it's definitely not a bagage here.

File: apm-agent-plugins/apm-apache-httpclient/apm-apache-httpclient3-plugin/src/main/java/co/elastic/apm/agent/httpclient/v3/HttpClient3Instrumentation.java [hunk 12]
```diff
@@ -82,54 +80,46 @@ public static class HttpClient3Advice {
...
-            try {
-                httpClientURI = httpMethod.getURI();
-                host = hostConfiguration.getHost();
-                port = hostConfiguration.getPort();
-                protocol = hostConfiguration.getProtocol().getScheme();
-                uri = httpClientURI.toString();
-            } catch (Exception e) {
+            Span<?> span = null;
+            final ElasticContext<?> parent = TracerAwareInstrumentation.tracer.currentContext();
+            if (parent.getAbstractSpan() != null) {
```
Comment (SylvainJuge) [inline apm-agent-plugins/apm-apache-httpclient/apm-apache-httpclient3-plugin/src/main/java/co/elastic/apm/agent/httpclient/v3/HttpClient3Instrumentation.java:85]: [nit] I would still prefer the shorcut to keep the code a bit more linear (but that's just a preference, maybe let's discuss this during weekly meeting).

File: apm-agent-plugins/apm-apache-httpclient/apm-apache-httpclient4-plugin/src/main/java/co/elastic/apm/agent/httpclient/v4/ApacheHttpAsyncClientInstrumentation.java [hunk 13]
```diff
@@ -89,41 +91,44 @@ public static class ApacheHttpAsyncClientAdvice {
...
-                    .withSync(false)
-                    .activate();
-
-                wrappedProducer = asyncHelper.wrapRequestProducer(requestProducer, span, null);
-                wrappedFutureCallback = asyncHelper.wrapFutureCallback(futureCallback, context, span);
-                responseFutureWrapped = true;
-            } else {
-                wrappedProducer = asyncHelper.wrapRequestProducer(requestProducer, null, parent);
+            Span<?> span = null;
+            ElasticContext<?> parent = tracer.currentContext();
```
Comment (SylvainJuge) [inline apm-agent-plugins/apm-apache-httpclient/apm-apache-httpclient4-plugin/src/main/java/co/elastic/apm/agent/httpclient/v4/ApacheHttpAsyncClientInstrumentation.java:103]: [minor] do you think that would make more sense to call this kind of variables `activeContext` instead of `parent` ? Since the context can now be something else than a parent in a parent/child relationship I think it would make that a bit more clear.

File: apm-agent-plugins/apm-apache-httpclient/apm-apache-httpclient4-plugin/src/main/java/co/elastic/apm/agent/httpclient/v4/ApacheHttpClientInstrumentation.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (SylvainJuge) [reply apm-agent-plugins/apm-apache-httpclient/apm-apache-httpclient4-plugin/src/main/java/co/elastic/apm/agent/httpclient/v4/ApacheHttpClientInstrumentation.java:62]: Sorry got confused by the `// re-adds the header on redirects` comment (which was removed), it completely makes sense to activate the exit span we've just created here.

File: apm-agent-plugins/apm-apache-httpclient/apm-apache-httpclient4-plugin/src/test/java/co/elastic/apm/agent/httpclient/v4/ApacheHttpAsyncClientInstrumentationTest.java [hunk 14]
```diff
@@ -80,4 +85,23 @@ public void cancelled() {
 
         responseFuture.get();
     }
+
+    @Test
+    public void testSpanFinishOnCallbackInvocation() throws Exception {
```
Comment (SylvainJuge) [inline apm-agent-plugins/apm-apache-httpclient/apm-apache-httpclient4-plugin/src/test/java/co/elastic/apm/agent/httpclient/v4/ApacheHttpAsyncClientInstrumentationTest.java:90]: :star_struck: nice addition !

File: apm-agent-plugins/apm-api-plugin/src/main/java/co/elastic/apm/agent/pluginapi/LegacySpanInstrumentation.java [hunk 15]
```diff
@@ -104,7 +104,14 @@ public static class AdviceClass {
             @Advice.OnMethodExit(inline = false)
             public static Object doCreateSpan(@Advice.FieldValue(value = "span", typing = Assigner.Typing.DYNAMIC) Object span) {
                 if (span instanceof Span<?>) {
-                    return ((Span<?>) span).createSpan();
+                    Span<?> spanTyped = (Span<?>) span;
+                    //TODO: make the public API work on contexts instead of spans, otherwise baggage might be lost
```
Comment (SylvainJuge) [inline apm-agent-plugins/apm-api-plugin/src/main/java/co/elastic/apm/agent/pluginapi/LegacySpanInstrumentation.java:108]: I'm not sure what you meant here, is that something that will be fixed by your next PR when adding baggage ?

File: apm-agent-plugins/apm-api-plugin/src/main/java/co/elastic/apm/agent/pluginapi/CaptureSpanInstrumentation.java [hunk 16]
```diff
@@ -85,8 +86,9 @@ public static Object onMethodEnter(
                 defaultValueProvider = AnnotationValueOffsetMappingFactory.TrueDefaultValueProvider.class
             ) boolean discardable
         ) {
-            final AbstractSpan<?> parent = tracer.getActive();
-            if (parent == null) {
+            ElasticContext<?> parent = tracer.currentContext();
+            final AbstractSpan<?> parentSpan = parent.getAbstractSpan();
```
Comment (SylvainJuge) [inline apm-agent-plugins/apm-api-plugin/src/main/java/co/elastic/apm/agent/pluginapi/CaptureSpanInstrumentation.java:90]:
```suggestion
            ElasticContext<?> currentContext = tracer.currentContext();
            final AbstractSpan<?> parentSpan = currentContext.getAbstractSpan();
```

File: apm-agent-plugins/apm-api-plugin/src/main/java/co/elastic/apm/agent/pluginapi/TracedInstrumentation.java [hunk 17]
```diff
@@ -92,8 +93,9 @@ public static Object onMethodEnter(
                 defaultValueProvider = AnnotationValueOffsetMappingFactory.TrueDefaultValueProvider.class
             ) boolean discardable) {
 
-            final AbstractSpan<?> parent = tracer.getActive();
-            if (parent != null) {
+            final ElasticContext<?> parent = tracer.currentContext();
+            final AbstractSpan<?> parentSpan = parent.getAbstractSpan();
+            if (parentSpan != null) {
                 if (parent.shouldSkipChildSpanCreation()) {
```
Comment (SylvainJuge) [inline apm-agent-plugins/apm-api-plugin/src/main/java/co/elastic/apm/agent/pluginapi/TracedInstrumentation.java:99]:
```suggestion
            final ElasticContext<?> currentContext = tracer.currentContext();
            final AbstractSpan<?> parentSpan = currentContext.getAbstractSpan();
            if (parentSpan != null) {
                if (currentContext.shouldSkipChildSpanCreation()) {
```

File: apm-agent-plugins/apm-asynchttpclient-plugin/src/main/java/co/elastic/apm/agent/asynchttpclient/AbstractAsyncHttpClientInstrumentation.java [hunk 18]
```diff
@@ -99,23 +100,22 @@ public static class AdviceClass {
             @Advice.OnMethodEnter(suppress = Throwable.class, inline = false)
             public static Object onBeforeExecute(@Advice.Argument(value = 0) Request request,
                                                  @Advice.Argument(value = 1) AsyncHandler<?> asyncHandler) {
-                final AbstractSpan<?> parent = tracer.getActive();
-                if (parent == null) {
-                    return null;
+                final ElasticContext<?> parent = tracer.currentContext();
+                final AbstractSpan<?> parentSpan = tracer.getActive();
```
Comment (SylvainJuge) [inline apm-agent-plugins/apm-asynchttpclient-plugin/src/main/java/co/elastic/apm/agent/asynchttpclient/AbstractAsyncHttpClientInstrumentation.java:104]:
```suggestion
                final ElasticContext<?> currentContext = tracer.currentContext();
                final AbstractSpan<?> parentSpan = currentContext.getActive();
```

File: apm-agent-plugins/apm-finagle-httpclient-plugin/src/main/java/co/elastic/apm/agent/finaglehttpclient/FinaglePayloadSizeFilterInstrumentation.java [hunk 19]
```diff
@@ -123,41 +123,35 @@ public static Object onBeforeExecute(@Nullable @Advice.Argument(0) Request reque
...
-
-            URI uri = resolveURI(request, host);
-            Span<?> span = HttpClientHelper.startHttpClientSpan(parent, request.method().name(), uri, null);
-
-            if (span != null) {
-                span.activate();
-                if (hostUnknown) {
-                    inflightSpansWithUnknownHost.put(request, span);
+            ElasticContext<?> parentContext = tracer.currentContext();
+            AbstractSpan<?> parent = parentContext.getAbstractSpan();
```
Comment (SylvainJuge) [inline apm-agent-plugins/apm-finagle-httpclient-plugin/src/main/java/co/elastic/apm/agent/finaglehttpclient/FinaglePayloadSizeFilterInstrumentation.java:127]:
```suggestion
            ElasticContext<?> currentContext = tracer.currentContext();
            AbstractSpan<?> parent = currentContext.getAbstractSpan();
```

File: apm-agent-plugins/apm-httpclient-core/src/main/java/co/elastic/apm/agent/httpclient/HttpClientHelper.java [hunk 20]
```diff
@@ -34,7 +34,7 @@ public class HttpClientHelper {
     public static final String HTTP_SUBTYPE = "http";
 
     @Nullable
-    public static Span<?> startHttpClientSpan(AbstractSpan<?> parent, String method, @Nullable URI uri, @Nullable CharSequence hostName) {
+    public static Span<?> startHttpClientSpan(ElasticContext<?> parent, String method, @Nullable URI uri, @Nullable CharSequence hostName) {
```
Comment (SylvainJuge) [inline apm-agent-plugins/apm-httpclient-core/src/main/java/co/elastic/apm/agent/httpclient/HttpClientHelper.java:37]:
```suggestion
    public static Span<?> startHttpClientSpan(ElasticContext<?> activeContext, String method, @Nullable URI uri, @Nullable CharSequence hostName) {
```

File: apm-agent-plugins/apm-httpclient-core/src/main/java/co/elastic/apm/agent/httpclient/HttpClientHelper.java [hunk 21]
```diff
@@ -50,8 +50,8 @@ public static Span<?> startHttpClientSpan(AbstractSpan<?> parent, String method,
     }
 
     @Nullable
-    public static Span<?> startHttpClientSpan(AbstractSpan<?> parent, String method, @Nullable String uri,
-                                           @Nullable String scheme, @Nullable CharSequence hostName, int port) {
+    public static Span<?> startHttpClientSpan(ElasticContext<?> parent, String method, @Nullable String uri,
```
Comment (SylvainJuge) [inline apm-agent-plugins/apm-httpclient-core/src/main/java/co/elastic/apm/agent/httpclient/HttpClientHelper.java:53]:
```suggestion
    public static Span<?> startHttpClientSpan(ElasticContext<?> activeContext, String method, @Nullable String uri,
```

File: apm-agent-plugins/apm-jdbc-plugin/src/main/java/co/elastic/apm/agent/jdbc/helper/JdbcHelper.java [hunk 22]
```diff
@@ -81,8 +82,8 @@ public String retrieveSqlForStatement(Object statement) {
 
 
     @Nullable
-    public Span<?> createJdbcSpan(@Nullable String sql, Object statement, @Nullable AbstractSpan<?> parent, boolean preparedStatement) {
-        if (!(statement instanceof Statement) || sql == null || isAlreadyMonitored(parent) || parent == null) {
+    public Span<?> createJdbcSpan(@Nullable String sql, Object statement, ElasticContext<?> parent, boolean preparedStatement) {
```
Comment (SylvainJuge) [inline apm-agent-plugins/apm-jdbc-plugin/src/main/java/co/elastic/apm/agent/jdbc/helper/JdbcHelper.java:85]:
```suggestion
    public Span<?> createJdbcSpan(@Nullable String sql, Object statement, ElasticContext<?> currentContext, boolean preparedStatement) {
```

File: apm-agent-plugins/apm-jms-plugin/apm-jms-plugin-base/src/main/java/co/elastic/apm/agent/jms/JmsMessageConsumerInstrumentation.java [hunk 23]
```diff
@@ -92,12 +93,13 @@ public static Object beforeReceive(@Advice.Origin Class<?> clazz,
                 AbstractSpan<?> createdSpan = null;
                 boolean createPollingTransaction = false;
                 boolean createPollingSpan = false;
-                final AbstractSpan<?> parent = tracer.getActive();
-                if (parent == null) {
+                final ElasticContext<?> parent = tracer.currentContext();
+                final AbstractSpan<?> parentSpan = parent.getAbstractSpan();
```
Comment (SylvainJuge) [inline apm-agent-plugins/apm-jms-plugin/apm-jms-plugin-base/src/main/java/co/elastic/apm/agent/jms/JmsMessageConsumerInstrumentation.java:97]:
```suggestion
                final ElasticContext<?> currentContext = tracer.currentContext();
                final AbstractSpan<?> parentSpan = currentContext.getAbstractSpan();
```

File: apm-agent-plugins/apm-kafka-plugin/apm-kafka-base-plugin/src/main/java/co/elastic/apm/agent/kafka/KafkaConsumerInstrumentation.java [hunk 24]
```diff
@@ -62,8 +63,8 @@ public static class KafkaPollEntryAdvice {
         @SuppressWarnings("unused")
         @Advice.OnMethodEnter(suppress = Throwable.class, inline = false)
         public static void pollStart() {
-
-            final AbstractSpan<?> activeSpan = tracer.getActive();
+            final ElasticContext<?> parent = tracer.currentContext();
+            final AbstractSpan<?> activeSpan = parent.getAbstractSpan();
```
Comment (SylvainJuge) [inline apm-agent-plugins/apm-kafka-plugin/apm-kafka-base-plugin/src/main/java/co/elastic/apm/agent/kafka/KafkaConsumerInstrumentation.java:67]:
```suggestion
            final ElasticContext<?> currentContext = tracer.currentContext();
            final AbstractSpan<?> activeSpan = currentContext.getAbstractSpan();
```

File: apm-agent-plugins/apm-kafka-plugin/apm-kafka-base-plugin/src/main/java/co/elastic/apm/agent/kafka/KafkaConsumerInstrumentation.java [hunk 25]
```diff
@@ -76,7 +77,7 @@ public static void pollStart() {
                 }
             }
 
-            Span<?> span = activeSpan.createExitSpan();
+            Span<?> span = parent.createExitSpan();
```
Comment (SylvainJuge) [inline apm-agent-plugins/apm-kafka-plugin/apm-kafka-base-plugin/src/main/java/co/elastic/apm/agent/kafka/KafkaConsumerInstrumentation.java:80]:
```suggestion
            Span<?> span = currentContext.createExitSpan();
```

File: apm-agent-plugins/apm-kafka-plugin/apm-kafka-headers-plugin/src/main/java/co/elastic/apm/agent/kafka/KafkaProducerHeadersInstrumentation.java [hunk 26]
```diff
@@ -71,21 +71,21 @@ public static Object[] beforeSend(@Advice.FieldValue("apiVersions") final ApiVer
...
+                    headersHelper.setOutgoingTraceContextHeaders(tracer.currentContext(), record);
                 } catch (final IllegalStateException e) {
                     // headers are in a read-only state
                     logger.debug("Failed to add header to Kafka record {}, probably to headers' read-only state.", record);
                 }
             }
 
+            if (span == null) {
+                return null;
+            }
```
Comment (SylvainJuge) [inline apm-agent-plugins/apm-kafka-plugin/apm-kafka-headers-plugin/src/main/java/co/elastic/apm/agent/kafka/KafkaProducerHeadersInstrumentation.java:88]: :tada: it looks like you've fixed a bug where the context propagation headers might not have been sent when no span was created (which is definitely a corner case).

File: apm-agent-plugins/apm-vertx/apm-vertx4-plugin/src/main/java/co/elastic/apm/agent/vertx/v4/webclient/HttpContextInstrumentation.java [hunk 27]
```diff
@@ -122,7 +122,13 @@ public static void sendRequest(@Advice.This HttpContext<?> httpContext, @Advice.
 
                 if (parentSpan != null) {
                     AbstractSpan<?> parent = (AbstractSpan<?>) parentSpan;
-                    webClientHelper.startSpan(parent, httpContext, request);
+                    //TODO: do actual context propagation instead of span propagation
```
Comment (SylvainJuge) [inline apm-agent-plugins/apm-vertx/apm-vertx4-plugin/src/main/java/co/elastic/apm/agent/vertx/v4/webclient/HttpContextInstrumentation.java:125]: is that something you plan to do in the second PR ?

**Review by SylvainJuge — APPROVED** [review]
LGTM, I'm mostly nit-picking about minor style things, congrats on this huge refactor !!

File: apm-agent-core/src/main/java/co/elastic/apm/agent/impl/EmptyElasticContext.java [hunk 28]
```diff
@@ -0,0 +1,65 @@
...
+ */
+package co.elastic.apm.agent.impl;
+
+import co.elastic.apm.agent.impl.transaction.AbstractSpan;
+import co.elastic.apm.agent.impl.transaction.ElasticContext;
+import co.elastic.apm.agent.tracer.Scope;
+
+import javax.annotation.Nullable;
+
+class EmptyElasticContext extends ElasticContext<EmptyElasticContext> {
```
[the remark this answers is not in the document]
Comment (JonasKunz) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/impl/EmptyElasticContext.java:27]: This is unfortunately not really possible via inheritance because the implementation variant of `ElasticContext` is now a class, so `EmptyElasticContext` can't inherit from both `NoopElasticContext` and `ElasticContext`. We could do it via composition instead, but this is not worth it here imo. In addition in a follow up PR this class wil lbecome even smaller: In the follow up PR I've moved the activation method implementations up into `ElasticContext`. The main reason is that `EmptyElasticContext.activate()` is a bit flawed with the current implementation: If a span/transaction/baggage is currently active and you invoke `EmptyElasticContext.activate()`, it doesn't "clear" the currently active span/transaction/baggage. This is nothing we do in production code, but it is useful in tests and otherwise also counterintuitive.

File: apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/AbstractSpan.java [hunk 29]
```diff
@@ -619,6 +617,14 @@ private boolean hasChildId(Id spanId) {
         return false;
     }
 
+    /**
+     * If this AbstractSpan is a transaction, then the AbstractSpan itself is returned.
+     * Otherwise, the parent transaction of the span is returned.
```
[the remark this answers is not in the document]
Comment (JonasKunz) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/AbstractSpan.java:622]: Yes, I'll use your wording. It's much clearer

File: apm-agent-plugins/apm-apache-httpclient/apm-apache-httpclient4-plugin/src/main/java/co/elastic/apm/agent/httpclient/v4/ApacheHttpAsyncClientInstrumentation.java (hunk 13 above)
Comment (JonasKunz) [reply apm-agent-plugins/apm-apache-httpclient/apm-apache-httpclient4-plugin/src/main/java/co/elastic/apm/agent/httpclient/v4/ApacheHttpAsyncClientInstrumentation.java:103]: I'm fine with both. The `ElasticContext` is now the one starting the span, so technically it is the parent. The created span inherits the parent-span and the baggage from it's parent context.
I'll rename it to `activeContext` though

File: apm-agent-plugins/apm-api-plugin/src/main/java/co/elastic/apm/agent/pluginapi/LegacySpanInstrumentation.java (hunk 15 above)
Comment (JonasKunz) [reply apm-agent-plugins/apm-api-plugin/src/main/java/co/elastic/apm/agent/pluginapi/LegacySpanInstrumentation.java:108]: I'll remove the "TODO" and replace it with a more clear comment as this limitation will stay.
The limitation is if you start a span on the public API via `ElasticAPM.currentSpan().startSpan()` then you will loose baggage updates which happened after the `currentSpan()` was started.
Imo it is not worth investing in lifting this limitation, as this can only occur if you are actively mixing Otel (for accessing baggage) and our own public API. The solution is to just stay with Otel.

I'll also document this limitation on the Baggage issue.

File: apm-agent-plugins/apm-vertx/apm-vertx4-plugin/src/main/java/co/elastic/apm/agent/vertx/v4/webclient/HttpContextInstrumentation.java (hunk 27 above)
Comment (JonasKunz) [reply apm-agent-plugins/apm-vertx/apm-vertx4-plugin/src/main/java/co/elastic/apm/agent/vertx/v4/webclient/HttpContextInstrumentation.java:125]: Yes. It won't be in the next PR, but in the one after where I adapt all of our intra-process context propagations.
I did this after the addition of baggage so that I could actually add tests for it.
